chia/full_node/full_node.py

```diff
@@ -1316,21 +1356,44 @@ async def add_block_batch(
             f"CLVM: {sum(pvr.timing/1000.0 for pvr in pre_validation_results):0.2f}s "
             f"({len(blocks_to_validate)} blocks, start height: {blocks_to_validate[0].height})",
         )
+
         for i, block in enumerate(blocks_to_validate):
             if pre_validation_results[i].error is not None:
-                self.log.error(f"Invalid block from peer: {peer_info} {Err(pre_validation_results[i].error)}")
-                return False, None, Err(pre_validation_results[i].error)
+                self.log.error(
+                    f"Invalid block from peer: {peer_info} height {block.height} {Err(pre_validation_results[i].error)}"
+                )
+                return (
+                    False,
+                    None,
+                    current_ssi,
+                    current_difficulty,
+                    prev_ses_block,
+                    Err(pre_validation_results[i].error),
+                )
 
         agg_state_change_summary: Optional[StateChangeSummary] = None
-
+        block_record = await self.blockchain.get_block_record_from_db(blocks_to_validate[0].prev_header_hash)
         for i, block in enumerate(blocks_to_validate):
             assert pre_validation_results[i].required_iters is not None
             state_change_summary: Optional[StateChangeSummary]
             # when adding blocks in batches, we won't have any overlapping
             # signatures with the mempool. There won't be any cache hits, so
             # there's no need to pass the BLS cache in
+
+            if len(block.finished_sub_slots) > 0:
+                cc_sub_slot = block.finished_sub_slots[0].challenge_chain
+                if cc_sub_slot.new_sub_slot_iters is not None or cc_sub_slot.new_difficulty is not None:
+                    expected_sub_slot_iters, expected_difficulty = get_next_sub_slot_iters_and_difficulty(
+                        self.constants, True, block_record, self.blockchain
+                    )
+                    assert cc_sub_slot.new_sub_slot_iters is not None
+                    current_ssi = cc_sub_slot.new_sub_slot_iters
+                    assert cc_sub_slot.new_difficulty is not None
+                    current_difficulty = cc_sub_slot.new_difficulty
+                    assert expected_sub_slot_iters == current_ssi
+                    assert expected_difficulty == current_difficulty
             result, error, state_change_summary = await self.blockchain.add_block(
-                block, pre_validation_results[i], None, fork_info
+                block, pre_validation_results[i], None, fork_info, current_ssi, prev_ses_block=prev_ses_block
```

Comment:
If one of the blocks in `all_blocks` is a ses block, and we end up updating `prev_ses_block` in the block at line 1305, we don't reset it back anywhere, even though here we start over the iteration again. It seems like it might be set wrong for the first blocks. Same thing goes for `current_ssi` and `current_difficulty`, as far as I can tell

Reply:
the update in 1305 if for blocks that are already validated and would not be iterated in the validation process as they would not be included in blocks_to_validate